Support pre-created floating IPs with Terraform in Openstack

anders-elastisys opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md and developer guide https://git.k8s.io/community/contributors/devel/development.md
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

This PR adds support for specifying existing floating IPs on each machine listed in `k8s_nodes` and `k8s_masters` with a new optional variable `reserved_floating_ip` which is only used when the `floating_ip` variable is set to true. If `reserved_floating_ip` is not set and `floating_ip` is true, a new floating IP will be created same as before.

When `reserved_floating_ip` is set, that IP will get released and not deleted when deleting all Terraform resources.

**Known limitations:**
- If you specify the same IP for `reserved_floating_ip` on multiple hosts, Terraform will not complain and only one machine gets assigned the IP in the end, as there is currently no check for whether or not the specific IP has already been assigned. The output will show the same IP several times as well.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/elastisys/compliantkubernetes-kubespray/issues/392

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note

```

Xartos commented on contrib/terraform/openstack/README.md:
```diff
@@ -317,7 +317,8 @@ k8s_nodes:
     az: string # Name of the AZ
     flavor: string # Flavor ID to use
-    floating_ip: bool # If floating IPs should be created or not
+    create_floating_ip: bool # If floating IPs should be created or not
+    reserved_floating_ip: string # (optional) Use existing floating IP
```
Suggestion: For backwards compatibility, should we do

```suggestion
    floating_ip: bool # If floating IPs should be used
    reserved_floating_ip: string # (optional) Use existing floating IP, if not set or set to "" a new IP will be created
```

So that the first variable only sets if an FIP should be used or not. And then the `reserved_floating_ip` dictates if a new FIP should be created or if an existing one should be used

anders-elastisys commented on contrib/terraform/openstack/modules/compute/main.tf:
```diff
@@ -506,13 +506,13 @@ resource "openstack_compute_instance_v2" "k8s_masters" {
   metadata = {
     ssh_user         = var.ssh_user
-    kubespray_groups = "%{if each.value.etcd == true}etcd,%{endif}kube_control_plane,${var.supplementary_master_groups},k8s_cluster%{if each.value.floating_ip == false},no_floating%{endif}"
+    kubespray_groups = "%{if each.value.etcd == true}etcd,%{endif}kube_control_plane,${var.supplementary_master_groups},k8s_cluster%{if each.value.create_floating_ip == false},no_floating%{endif}"
```
Right, yes it should be an additional check here for the kubespray groups, will fix!

anders-elastisys commented:
@Xartos PTAL, I reverted the name change for `floating_ip`, changed the logic, fixed some missing checks and also fixed the output for floating ips so that the reserved ones are included as well. I had to change the `reserved_floating_ip` to not be optional anymore as I wasn't sure how to handle it in the `ips` module as I got an error when checking if it was set to null since ig the variable is not really declared in that module: `This object does not have an attribute named "reserved_floating_ip".`

Xartos commented on contrib/terraform/openstack/modules/ips/main.tf:
```diff
@@ -15,7 +15,7 @@ resource "openstack_networking_floatingip_v2" "k8s_master" {
 
 resource "openstack_networking_floatingip_v2" "k8s_masters" {
-  for_each   = var.number_of_k8s_masters == 0 && var.number_of_k8s_masters_no_etcd == 0 ? { for key, value in var.k8s_masters : key => value if value.floating_ip } : {}
+  for_each   = var.number_of_k8s_masters == 0 && var.number_of_k8s_masters_no_etcd == 0 ? { for key, value in var.k8s_masters : key => value if value.floating_ip && (value.reserved_floating_ip == "") } : {}
```
Was this the one that gave you trouble when `reserved_floating_ip` is set to optional? I assume this was `value.reserved_floating_ip == null` ?

If so it's very strange that it errored out :thinking: It shouldn't be an issue.

anders-elastisys replied:
Yes, here is the full error if I try with the `null` check and don't set the variable for one of the `k8s_masters`:
```console
╷
│ Error: Unsupported attribute
│ 
│   on modules/ips/main.tf line 18, in resource "openstack_networking_floatingip_v2" "k8s_masters":
│   18:   for_each   = var.number_of_k8s_masters == 0 && var.number_of_k8s_masters_no_etcd == 0 ? { for key, value in var.k8s_masters : key => value if value.floating_ip && (value.reserved_floating_ip == null) } : {}
│ 
│ This object does not have an attribute named "reserved_floating_ip".
```

Xartos replied:
Aah, right probably because it's in the for loop. 
But you should be able to use the [lookup function](https://developer.hashicorp.com/terraform/language/functions/lookup) to solve that. something like

`lookup(value, "reserved_floating_ip", "") == ""`

Xartos commented on contrib/terraform/openstack/modules/compute/outputs.tf:
```diff
@@ -1,3 +1,11 @@
   value = concat(openstack_compute_instance_v2.k8s_master_no_floating_ip.*, openstack_compute_instance_v2.k8s_master_no_floating_ip_no_etcd.*)
 }
+
+output "k8s_masters_reserved_ips" {
```
Wouldn't it be possible to have this in the `k8s_master_fips` variable in "outputs.tf" in the "ips" module instead?

I'm just thinking that then any script or such that assumes the FIP is there would still work even with reserved IPs

But if you think that is too much trouble then I'm fine leaving it here

anders-elastisys replied:
I think I looked at doing it that way but encountered some issues, but I can try looking at it again. I agree it would be nicer

anders-elastisys replied:
I am not sure how I would go about it, since the output for `openstack_networking_floatingip_v2.k8s_masters` is a map with a bunch of [attributes](https://registry.terraform.io/providers/terraform-provider-openstack/openstack/latest/docs/resources/networking_floatingip_v2#argument-reference), I am not sure how to combine it with the reserved_ips. I tried by creating a local object for `reserved_ips` that have the `address` attribute, but then it complains that not all map elements/attributes are the same.

Merging maps for both reserved and created floating IPs with only the `address` attribute could be a solution, see https://github.com/elastisys/kubespray/pull/19/commits/2b80ad636dbbc83d0fb46526d8ab88bbb4061bbc for how this could look like, in this case only for `k8s_masters`. If this solution suffices I will update it for `k8s_nodes` as well.

Xartos replied:
I think that looks like a good solution. As long as we can assure that stuff other than the address is used anywhere so we don't break anything.

OlleLarsson commented on contrib/terraform/openstack/README.md:
```diff
@@ -299,10 +299,10 @@ For your cluster, edit `inventory/$CLUSTER/cluster.tfvars`.
-| `k8s_master_loadbalancer_public_ip` | Specify if an existing floating IP should be used for the load balancer. A new floating IP is assigned by default  |
+|`k8s_master_loadbalancer_enabled` | Enable and use an Octavia load balancer for the K8s master nodes |
+|`k8s_master_loadbalancer_listener_port` | Define via which port the K8s Api should be exposed. `6443` by default  |
+|`k8s_master_loadbalancer_server_port` | Define via which port the K8S api is available on the mas. `6443` by default |
```
```suggestion
|`k8s_master_loadbalancer_server_port` | Define via which port the K8S api is available on the master. `6443` by default |
```

OlleLarsson reviewed (approved):
LGTM after seeing a live demo :raised_hands:

anders-elastisys commented:
I tested running kubespray with this change and a dynamic inventory file, with some nodes having `reserved_floating_ip` set and some nodes creating new ones and I did not encounter any issues, kubespray ran successfully.

Xartos reviewed (approved):
Nice!
